### ISSUE TITLE
Delete files as media items are deleted

### ISSUE THREAD

**Issue by willguv** [localgovdrupal/localgov_core#252]
Mentioned at the PG on 25th September

As media items are deleted, the attached files are left behind - this is causing a problem for a few councils

@andybroomfield identified 'media file delete' which might be in the default LGD install already

![Image](https://github.com/user-attachments/assets/48b25aff-906f-4c7f-aa84-463c9a7586c9)

_Originally posted by @andybroomfield in https://github.com/localgovdrupal/localgov_core/issues/220#issuecomment-2247942737_

**Comment by mccrodp** [comment]
Hey folks (cc @willguv). Just looking to understand this issue as it's not clear the intention yet. I read the linked thread and will take a stab at it:

# Problem / Solution
## Problem
* As media items are deleted, the attached files are left behind - this is causing a problem for a few councils.

## Context
* The default behaviour in Drupal is for files attached to deleted entities to be flagged for removal "later" and hence are kept temporarily in the interim.

## Outcome
* Some councils may want this default behaviour (see context) and some would like to delete files instantly.

## Solution
???

### Questions
1. How far away is "later" for when files are actually deleted? Is it eg. on next cron run?
1. There was a mention in the other thread of a "setting for how long temporary files are kept", any details on that?
2. Do councils that want instant delete of attached files, want that to be a global setting, or a setting per deleted entity?

There was a suggestion to use https://www.drupal.org/project/media_file_delete if it was already in use in LGD, but I don't see it in `contrib` folder. Also, that adds an adds an additional module dependency which may not be what councils are looking for (see Q3 above). The suggested module seems to provide a per entity setting to delete the attached files instantly, rather than a global setting.

Please correct any of the above where I have misunderstood things. I also gathered that some councils may want to default behaviour, so please confirm that. If every council just wanted instant delete of files attached to deleted media entities, it'd be simpler to just use `hook_entity_delete` and `if ($file) $file->delete()`, but I presume that's not the hardline solution that all councils want 🤔

**Comment by willguv** [comment]
Hi @mccrodp thanks for this detail. 

What I've heard from councils is that "old files are left lying around taking up space" or words to that effect. I'm not clear on how often this happens, or by what cause. If the default behavior eventually clears away old files, then it feels like there's not much of a problem.

If this prompts more thoughts, please do post them. I think I need to understand the issue a little better!

**Comment by mccrodp** [comment]
>If the default behavior eventually clears away old files, then it feels like there's not much of a problem.

Right, that is getting us/me closer to something I think, thanks @willguv. So, if I understand the actions needed then, they are:

1. Surface / detail the default behaviour so that everyone has a clear understanding of files associated to deleted media entities
1. Get to the bottom of whether the default behaviour clears away these old files eventually, specifically highlighting: 
-- what conditions trigger this clearing
-- how often it happens
-- if there's anyway to trigger it or make it happen more regularly

~~I can take a closer look at this tomorrow~~ Didn't get time for this.

**Comment by MariosORION** [comment]
@mccrodp @willguv I've spent some time looking into this and here are my findings:

How Drupal handles files by default (also see https://www.drupal.org/docs/administering-a-drupal-site/uploaded-file-management). The key points are:

- Files will be saved as "permanent" by default. It is possible that some Councils might have implemented custom logic that leads to files marked as "temporary" present in the codebase.
- Temporary files get deleted (on the first Cron run) after 6 hours by default. This is a configuration setting under /admin/config/media/file-system and Councils will be able to change this if they need more/less time before temporary files get permanently deleted.
- [Media file delete](https://www.drupal.org/project/media_file_delete) acts upon the delete action of Media entities and gives editors the ability to delete the accompanying file immediately.
- [File Delete](https://www.drupal.org/project/file_delete) offers the ability to add a "Delete file" option on the Files screen (this has to be added to the related View's configuration). These files will be turned into temporary files and deleted after 6 hours (by default as described above) on a Cron run. There is also a "Delete files" permission this module adds that Councils can use based on how their roles are configured.

The File Delete module seems to be closer to what you describe @willguv.
Hope that helps @mccrodp ?

**Comment by finnlewis** [comment]
Discussing at Brighton Connect.... 

This is still a hot topic!

Justin at Southwark has the problem trying to delete a file but: 
 -> needs to delete from the media library
-> delete the file from the files tab
-> also clear the CDN cache.

So can we have one form interaction to rule them all?

"Delete this document... 

- [ ]   and the file? 
- [ ] and flush the CDN?

Other relevant issues: 

 - https://github.com/localgovdrupal/localgov_core/issues/294 
 - https://github.com/localgovdrupal/localgov_core/issues/179
 - https://github.com/localgovdrupal/localgov_core/pull/254
 - https://github.com/localgovdrupal/localgov_core/issues/162

Next steps: 

- [ ] Discuss at Product Drop-in (@willguv )
- [ ] Include interested folk from Southwark (Justin), Newcastle (Chris), Big Blue Door,  
- [ ] Define the acceptance test for something functionally better!
